manager/orchestrator/replicated/restart_test.go

```diff
@@ -803,3 +804,217 @@ func TestOrchestratorRestartWindow(t *testing.T) {
+	assert.Equal(t, Backoff2.Factor, gogotypes.DurationProto(factorTime))
+	assert.Equal(t, Backoff2.Max, gogotypes.DurationProto(maxTime))
+
+	// We failed once
+	assert.Equal(t, RestartSV1.GetFailuresSinceSuccess(observedTask2), uint64(1))
+
+	testutils.Expect(t, watch, state.EventCommit{})
+
+	delay2a := baseTime + factorTime
+	observedTask2a := testutils.WatchTaskUpdateDelay(t, watch, delay2a)
+	assert.Equal(t, observedTask2a.DesiredState, api.TaskStateRunning)
+	assert.Equal(t, observedTask2a.ServiceAnnotations.Name, "name1")
+}
+
+func TestOrchestratorTaskRestartDelay(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+	s := store.NewMemoryStore(nil)
+	assert.NotNil(t, s)
+	defer s.Close()
+
+	orchestrator := NewReplicatedOrchestrator(s)
+	defer orchestrator.Stop()
+
+	watch, cancel := state.Watch(s.WatchQueue() /*api.EventCreateTask{}, api.EventUpdateTask{}*/)
+	defer cancel()
+
+	baseTime := 10 * time.Millisecond
+	factorTime := 20 * time.Millisecond
+	maxTime := 4 * time.Second
+
+	// Create a service with two instances specified before the orchestrator is
+	// started. This should result in two tasks when the orchestrator
+	// starts up.
+	j1 := &api.Service{
+		ID: "id1",
+		Spec: api.ServiceSpec{
+			Annotations: api.Annotations{
+				Name: "name1",
+			},
+			Task: api.TaskSpec{
+				Runtime: &api.TaskSpec_Container{
+					Container: &api.ContainerSpec{},
+				},
+				Restart: &api.RestartPolicy{
+					Condition: api.RestartOnAny,
+					Backoff: &api.BackoffPolicy{
+						Base:   gogotypes.DurationProto(baseTime),
+						Factor: gogotypes.DurationProto(factorTime),
+						Max:    gogotypes.DurationProto(maxTime),
+					},
+					MaxAttempts: 2,
+				},
+			},
+			Mode: &api.ServiceSpec_Replicated{
+				Replicated: &api.ReplicatedService{
+					Replicas: 1,
+				},
+			},
+		},
+	}
+	err := s.Update(func(tx store.Tx) error {
+		assert.NoError(t, store.CreateService(tx, j1))
+		return nil
+	})
+	assert.NoError(t, err)
+
+	// Start the orchestrator.
+	go func() {
+		assert.NoError(t, orchestrator.Run(ctx))
+	}()
+
+	observedTask1 := testutils.WatchTaskCreate(t, watch)
+	assert.Equal(t, observedTask1.Status.State, api.TaskStateNew)
+	assert.Equal(t, observedTask1.ServiceAnnotations.Name, "name1")
+
+	// Check that the task has the correct BackoffPolicy values
+	Backoff1 := observedTask1.Spec.Restart.Backoff
+	assert.Equal(t, Backoff1.Base, gogotypes.DurationProto(baseTime))
+	assert.Equal(t, Backoff1.Factor, gogotypes.DurationProto(factorTime))
+	assert.Equal(t, Backoff1.Max, gogotypes.DurationProto(maxTime))
+
+	// Since observedTask1 hasn't failed yet, check that failuresAfterSuccess is 0
+	RestartSV1 := orchestrator.restarts
+	delayP, err := RestartSV1.TaskRestartDelay(ctx, observedTask1)
+	assert.NoError(t, err)
+
+	// Check that the delay duration is between 0 and the calculated backoff duration
+	// assert.Equal(t, *delayP, baseTime + factorTime)
+	assert.True(t, (0 <= *delayP) && (*delayP < baseTime+factorTime))
```

Comment:
If your feature doesn't work right, this test might still always succeed. For example, say the calculation fails and  `delayP == 0` always. This assertion will still pass.

That's sort of the problem with randomness, right? You can't really verify without statistical analysis.

It might be to refactor `TaskRestartDelay` such that you return the raw delay value and let the caller apply the random range. That would make it more testable. But maybe there is a better way to refactor it. I definitely don't think statistically analyzing the return value of the function in the test is the right approach tho, FWIW. 

This is not a blocker for me. Do what you feel is best.